### ISSUE TITLE
Split cache system into global, user and session namespace

### ISSUE THREAD

**Issue by lcharette** [userfrosting/UserFrosting#649]
To effectively fix #633, the cache system needs to be able to assign vars to a `global`, `user` and `session` namespace. This could also be useful for application that want's to store per user data (for example, cache each user friends list).

This could easily be factored into a facade and split into the appropriate UF core services. For example:
```
    $cacheService->forever("test", "1234");
    $cacheService->forever("foo", "bar");

    $currentUser->cache->forever("test", "4321");
    $currentUser->cache->forever("foo", "BARRRRRRRR");

    echo $cacheService->get('foo'); //Return "bar"
```

Problem is, [Laravel cache](https://laravel.com/docs/5.4/cache) doesn't allow to create two cache instances. 

A way around this would be to inject prefix, something like this :
```
    $cacheService->forever("_global.test", "1234");
    $cacheService->forever("_global.foo", "bar");

    $cacheService->forever("_users.2419.test", "4321");
    $cacheService->forever("_users.2419.foo", "BARRRRRRRR");

    echo $cacheService->get('_global.foo'); //Return "bar"
```

The problem with the above code is it would be impossible to flush a namespace cache, for example `_users.2419.*`, as Laravel cache doesn't include such a feature. 

This could easily be fixed using the [Laravel Cache `tag` feature](https://laravel.com/docs/5.4/cache#cache-tags). Unfortunately, `tags` are not supported by the file store, which is the default store for UserFrosting. [See this link as to why](http://tinyurl.com/j7cn3j4). Since I don't think making `Memcache` a requirement for UF (It will be recommended for production use once I write the doc), the only way around would be to extend the different cache function (`put`, `forever`, etc.) and keep a list (in cache) of the keys contained inside the namespace so they can be deleted one by one in a loop...

Or, is the ability to flush a 'namespace' really useful?

**Comment by lcharette** [comment]
Another idea would be to use the `tags` feature when available (so everything except `file` and `database` store) and provide fallback for the other two. 

For `file`, the cache system currently saves everything to `app/cache/illuminate`. I guess it would be possible to store files in different location:
```
app/cache/global/
app/cache/user2419/
app/cache/user9532/
app/cache/session12345667890/
```

For `database`....???

This would still requires a major facade for the Laravel cache manager and Stores...

**Comment by alexweissman** [comment]
A couple thoughts:

1. Really?  You can only define a single cache service?
2. What about caching to the DB by default?  DB is already a requirement for UF anyway.
3. Can you cache objects/arrays?

**Comment by alexweissman** [comment]
Oh, I see.  `database` won't support tags either:

> Cache tags are not supported when using the file or database cache drivers.

**Comment by alexweissman** [comment]
Looking at Laravel's config file for `cache`, it seems that this is already supported through the ability to define multiple stores: https://github.com/laravel/laravel/blob/master/config/cache.php#L20-L29

We don't necessarily need to re-implement Laravel's `Cache` facade (though it's not that hard since we have our own `Facade` base class already).  We could, for example, just create distinct services: `cache`, `userCache`, `sessionCache`, for example.

**Comment by lcharette** [comment]
Problem is the file Store send everything to the same directory in the file system. So multiple stores will overlap each other. I test that out yesterday. 

We can store arrays and object, even collections. Problem is their content will have the same expiration. 

Using `_global` and `_user` "namespace" is still the easiest way I found. But since we can't rely on user to use those, some kind of facade would be best I think. Probably something that could go in a new `userFrosting/cache` composer package

**Comment by lcharette** [comment]
(Façade being needed since we have to keep a map of key under that namespace so the whole namespace can be deleted st once)

**Comment by alexweissman** [comment]
You can't just do:

```
        'global' => [
            'driver' => 'file',
            'path' => storage_path('cache://global/'),
        ],
        'user' => [
            'driver' => 'file',
            'path' => storage_path('cache://user/'),
        ]
```

?

Can't we reverse-engineer what Laravel does to create these stores?

**Comment by lcharette** [comment]
I'm. It sure this is how it works, I'll have to test it out. 

We also have to keep in mind that the file driver is not intended for production use. A memcache or redis instance should be used in production (and discussed eventually in the deploy/cache portion of our doc). So the solution needs to works with every cache driver.

**Comment by lcharette** [comment]
I did a quick test and using multiple stores does work for the file driver. I'll have to test the memcache and redis one as I don't think you can just specify a different directory like you do with file. But, that opens a door as a simpler workaround for the file one. Memcache And redis can probably make use of tags or even maybe prefix.

**Comment by lcharette** [comment]
I did some research and tests and tags would indeed work for memcache type drivers while still allowing dev to use tags if they want to. 

I wrote a small class that create the namespaced cache instance using a common method for all drivers. I was thinking to post it as a new `UserFrosting/Cache` module as this would make it easier to include unit tests. 

Would that be too much of a breaking change, considering the cache configs in `core`will probably be simplified too?

**Comment by lcharette** [comment]
See first draft here : https://github.com/userfrosting/cache

I'll work at implementing this and #633 in a new `feature-cache` branch of the UF repo.

**Comment by lcharette** [comment]
This is now implemented in 4.1 !

Session cache is not directly accessible in the session service for now, as it's mostly used for alertStream (see #633). If there's a need for it in the future, it shouldn't be hard to add. User cache is what matters most anyway.
